src/Data/Array/Mutable/Linear.hs

```diff
@@ -79,18 +77,25 @@ data Array a where
 -- The size must be non-negative, otherwise this errors.
 alloc :: HasCallStack =>
   Int -> a -> (Array a #-> Ur b) #-> Ur b
-alloc size' x f
-  | size' >= 0 = f (Array size' (Unsafe.newMutArr size' x))
-  | otherwise =
-    (error ("Trying to allocate an array of size " ++ show size') :: x #-> x)
+alloc s x f
+  | s < 0 =
+    (error ("Array.alloc: negative size: " ++ show s) :: x #-> x)
     (f undefined)
+  | otherwise = Unlifted.alloc s x (\arr -> f (Array arr))
 
 -- | Allocate a constant array given a size and an initial value,
 -- using another array as a uniqueness proof.
 allocBeside :: Int -> a -> Array b #-> (Array b, Array a)
-allocBeside size val orig
-  | size >= 0 = (orig, Array size (Unsafe.newMutArr size val))
-  | otherwise = orig `lseq` error ("Trying to allocate an array of size " ++ show size)
+allocBeside s x (Array orig)
+  | s < 0 =
+     Unlifted.lseq
+       orig
+       (error ("Array.allocBeside: negative size: " ++ show s))
```

Comment:
I'm curious: doesn't `error (…) orig` work here? Is `error`'s type not polymorphic enough in the type representation?

Reply:
I think `error`'s type is polymorphic enough; but our problem is with linearity. Since type inference doesn't infer linear arrows (bold statement, let me know if this is incorrect), we need to put an explicit type signature. However, in this specific case it's especially annoying since `orig` is unlifted, so we should also specify the kind on that signature.

In the end, if we enable `-XKindSignatures` and `-XDataKinds`, we can do something like:

 ```haskell
    (error ("Array.allocBeside: negative size: " ++ show s)
      :: (a :: TYPE UnliftedRep) #-> b
    ) orig
```

I can not think a shorter way to express this, so in this state it's not worth the hassle.

Reply:
Fair enough.

Reply:
An easy trick is `error "something" $ linearVar1 linearVar2 ...` with the linear variant of `($)`. That works.

Reply:
Oh, that is smart. Thanks @Divesh-Otwani !